qi-lib/flow/core/compiler/deforest/fusion.rkt

```diff
@@ -0,0 +1,76 @@
+#lang racket/base
+
+(provide define-and-register-deforest-pass)
+
+(require (for-syntax racket/base
+                     syntax/parse)
+         syntax/parse
+         "syntax.rkt"
+         "../../passes.rkt"
+         "../../strategy.rkt")
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; The actual fusion generator implementation
+
+;; Used only in deforest-rewrite to properly recognize the end of
+;; fusable sequence.
+(define-syntax-class non-fusable
+  (pattern (~not (~or _:fst-syntax
+                      _:fsp-syntax
+                      _:fsc-syntax))))
+
+(define (make-deforest-rewrite generate-fused-operation)
+  (lambda (stx)
+    (syntax-parse stx
+      [((~datum thread) _0:non-fusable ...
+                        p:fsp-syntax
+                        ;; There can be zero transformers here:
+                        t:fst-syntax ...
+                        c:fsc-syntax
+                        _1 ...)
+       #:with fused (generate-fused-operation
+                     (syntax->list #'(p t ... c))
+                     stx)
+       #'(thread _0 ... fused _1 ...)]
+      [((~datum thread) _0:non-fusable ...
+                        t1:fst-syntax0
+                        t:fst-syntax ...
+                        c:fsc-syntax
+                        _1 ...)
+       #:with fused (generate-fused-operation
+                     (syntax->list #'(list->cstream t1 t ... c))
+                     stx)
+       #'(thread _0 ... fused _1 ...)]
+      [((~datum thread) _0:non-fusable ...
+                        p:fsp-syntax
+                        ;; Must be 1 or more transformers here:
+                        t:fst-syntax ...+
+                        _1 ...)
+       #:with fused (generate-fused-operation
+                     (syntax->list #'(p t ... cstream->list))
+                     stx)
+       #'(thread _0 ... fused _1 ...)]
+      [((~datum thread) _0:non-fusable ...
+                        f1:fst-syntax0
+                        f:fst-syntax ...+
+                        _1 ...)
+       #:with fused (generate-fused-operation
+                     (syntax->list #'(list->cstream f1 f ... cstream->list))
+                     stx)
+       #'(thread _0 ... fused _1 ...)]
+      ;; return the input syntax unchanged if no rules
+      ;; are applicable
+      [_ stx])))
+
+;; This syntax is actively used only once as it is intended to be used
+;; by alternative implementations. Currently only the CPS
+;; implementation uses it, however in the near future the named-let
+;; implementation will use it as well.
```

Comment:
This comment belongs in a commit message body, in my opinion. (Since it will be
with whatever commit introduces the macro and related changes, it will forever
make sense instead of going out of date if we add implementations and forget to
update the comment.)

Reply:
As agreed during our meeting, this information is useful in both the commit message and within the code as a comment.